### DISCUSSION TITLE
third-way extension: packcircle example when a ggplot2 dependancy or writing a spin off package doesn't feel right

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#54] [General]
With [packcircles/issues/3](https://github.com/mbedward/packcircles/issues/3#issuecomment-928588491) in mind, which @mbedward alerted me to.  It's a request for geom_packcircles() function a couple of years ago.

I could see something like this in packcircles documentation or just in the answer to this issue (kind of thinking about a world where ggcirclepack doesn't exist)...  

"Now that you've seen how to create a circle packing graph w/ ggplot2 by prepping your data and then feeding packed circle data into ggplot(), you might be interested in creating a convenience layer, geom_packcircles().  The prep is a little longer than just prepping the data to feed in, but you might find geom_packcircles() more intuitive to use if you are familiar with that framework"  

... w/ code below

I don't know if we're in the too much additional code for too little pay off zone, but maybe showing the single facet v continent-wrapped facet demonstrates the code saving).  Maybe this approach it could also be a kind of start point for thinking about a package or additional functionality.  

``` r
library(tidyverse)
library(packcircles)
compute_panel_circlepack <- function(data, scales){

  data_w_id <- data |>
    mutate(id = row_number())

  data_w_id |>
    pull(area) |>
    packcircles::circleProgressiveLayout(
      sizetype = 'area') |>
    packcircles::circleLayoutVertices(npoints = 50) |>
    left_join(data_w_id) |>
    mutate(group = id)

}

geom_packcircles <- function(...){

  StatTemp <- ggproto("StatTemp", Stat, 
                      compute_panel = compute_panel_circlepack)
  
  geom_polygon(stat = StatTemp, ...)

}

gapminder::gapminder |> 
  filter(year == 2002) |> 
  ggplot() + 
  aes(area = pop/1000000) + 
  geom_packcircles() + 
  aes(fill = continent) 
#> Joining with `by = join_by(id)`
```

![](https://i.imgur.com/yTHCJae.png)

``` r

last_plot() +
  facet_wrap(facet = vars(continent)) + 
  coord_equal()
#> Joining with `by = join_by(id)`
#> Joining with `by = join_by(id)`
#> Joining with `by = join_by(id)`
#> Joining with `by = join_by(id)`
#> Joining with `by = join_by(id)`
```

![](https://i.imgur.com/wmW7CX7.png)

<sup>Created on 2024-10-30 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by mbedward** [comment]
This looks really great. It would nice to add it to the packcircles docs and I'm sure users would appreciate it, but are you suggesting that this would be in lieu of a separate ggcirclepack package?

**Reply by EvaMaeRey** [reply]
Not planning to abandon ggcirclepack - just wanting to think about the circumstance of finding ggplot2 extension outside of the scope of core functionality of a package but not wanting to necessarily support a spin-off package. While recognizing the value of providing a ggplot2 layer that does some nice compute under the hood!  I'm not sure to what extent this is being done in other packages, but thought it might be interesting to explore.

**Reply by EvaMaeRey** [reply]
todo/think about: new `make_constructor` in ggplot2 v.4.0.0 could make the geom_ even more 'true' to the ggplotting experience.   And text layer would need to be demoed as well.

**Reply by EvaMaeRey** [reply]
rewriting w/ make_constructor

```r
library(tidyverse)
library(packcircles)

# wrap up compute you might do anyway
compute_panel_circlepack <- function(data, scales){

  data_w_id <- data |>
    mutate(id = row_number())

  data_w_id |>
    pull(area) |>
    packcircles::circleProgressiveLayout(sizetype = 'area') |>
    packcircles::circleLayoutVertices(npoints = 50) |>
    left_join(data_w_id) |>
    mutate(group = id)

}

# write Stat and geom_* user facing function
StatCirclePack <- ggproto("StatCirclePack", Stat, compute_panel = compute_panel_circlepack)
geom_packcircles <- make_constructor(GeomPolygon, stat = StatCirclePack)
```

### Enjoy...

```r
gapminder::gapminder |> 
  filter(year == 2002) |> 
  ggplot() + 
  aes(area = pop/1000000) + 
  geom_packcircles() + 
  aes(fill = continent) 
```

**Reply by EvaMaeRey** [reply]
@mbedward, if you are still interested, 🙃 where do you see this going in the docs?   With ggplot2 v4.0, we have the handy make_constructor function that makes writing the geom_packcircles function concise and correct!  What I'd suggested in at first was a little non-standard.

**Comment by EvaMaeRey** [comment]
Another place where this demo-in-documentation strategy might be of interest is geo-data packages.  The main purpose is to deliver data.  But maybe they could 'try on' extension with a demo of alternate API, perhaps adding to package functionality later or spinning off to gg* package later.   So, StatBrasilestados.  And then feed flat data in to ggplot()

Below, mostly reducing code from https://github.com/EvaMaeRey/sf2stat, which tries to deliver ready to go geom_region() interface.

compute_panel_estado <- function(data, scales){
  
   brasil_geo_estado %>% # includes geometry column
   ggplot2::StatSf$compute_panel(coord = ggplot2::CoordSf) |>
   ggplot2::StatSfCoordinates$compute_group(coord = ggplot2::CoordSf) |>
   inner_join(data)
    
}

StatBrasilestados <- ggproto("StatSfJoin", Stat, 
                        compute_panel = compute_panel_estados, 
                        default_aes = aes(label = after_stat(estado_name)),
                        required_aes = c("estado_nome|estado_codigo"))

estados_caracteristica |>  # two var data - no geometry column with borders
ggplot() + 
aes(estado_name = estado_var) + 
geom_sf(stat = StatBrasilestados) + 
geom_label(stat = StatBrasilestados) + 
aes(fill = population)

And if you will use a bunch, then create in-script convenience functions...

geom_estado <- function(  ...){geom_sf(stat = StatBrasilestados,...)}
geom_estado_label <- function(  ...){geom_label(stat = StatBrasilestados,...)}

**Reply by EvaMaeRey** [reply]
Mostly there, but forgot that coords come from plot data and won't automatically be retrieved from Stat.  So it's not quite as simple - have to add crs each layer.

``` r
library(tidyverse)


compute_panel_state <- function(data, scales){

usmapdata::us_map() %>% 
    rename(geometry = geom) %>% # includes geometry column
    ggplot2::StatSf$compute_panel(coord = ggplot2::CoordSf) |>
    ggplot2::StatSfCoordinates$compute_group(coord = ggplot2::CoordSf) |>
    inner_join(data)

}

StatUSstates <- ggproto("StatUSstates", 
                             Stat,
                             compute_panel = compute_panel_state,
                             default_aes = aes(label = after_stat(full)),
                             required_aes = c("full|abb|fips"))

USArrests |>
  rownames_to_column("state") |># two var data - no geometry column with borders
  ggplot() +
  aes(full = state) +
  geom_sf(stat = StatUSstates) +
  geom_sf_text(stat = StatUSstates, check_overlap = T) +
  aes(fill = Murder)
#> Warning in layer_sf(data = data, mapping = mapping, stat = stat, geom =
#> GeomText, : Ignoring unknown parameters: `fun.geometry`
#> Joining with `by = join_by(full)`
#> Joining with `by = join_by(full)`
```

![](https://i.imgur.com/dnIEtbE.png)

``` r

#And if you will use a bunch, then create in-script convenience functions...

geom_state <- function( ...){geom_sf(stat = StatUSstates,...)}
geom_state_text <- function( ...){geom_sf_text(stat = StatUSstates,...)}
geom_state_label <- function( ...){geom_sf_label(stat = StatUSstates,...)}

USArrests |>
  rownames_to_column("state") |># two var data - no geometry column with borders
  ggplot() +
  aes(full = state) +
  geom_state() + 
  geom_state_text() + 
  aes(fill = Murder) + 
  aes(label = Murder) + 
  scale_fill_viridis_c()
#> Warning in layer_sf(data = data, mapping = mapping, stat = stat, geom =
#> GeomText, : Ignoring unknown parameters: `fun.geometry`
#> Joining with `by = join_by(full)`
#> Joining with `by = join_by(full)`
```

![](https://i.imgur.com/joXOlSr.png)

<sup>Created on 2024-11-14 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
This fixes the second part..

usmapcrs <- sf::st_crs(usmapdata::us_map())

geom_state <- function( ...){c(geom_sf(stat = StatUSstates, ...), 
                               coord_sf(crs = usmapcrs))}
geom_state_text <- function( ...){c(geom_sf_text(stat = StatUSstates,...),
                                    coord_sf(crs = usmapcrs))}
geom_state_label <- function( ...){c(geom_sf_label(stat = StatUSstates,...),
                                   coord_sf(crs = usmapcrs))}

**Comment by EvaMaeRey** [comment]
Just quickly worked through an Australia example.

``` r
library(tidyverse)
library(ozmaps)
sf_oz <- ozmap("states")
```

![](https://i.imgur.com/NqnhQTC.png)

``` r
crs_au_states <- sf::st_crs(sf_oz)

sf_oz |> names()
#> [1] "NAME"     "geometry"
```

``` r
sf_oz |>
  select(state_name = NAME) %>% 
    ggplot2::StatSf$compute_panel(coord = ggplot2::CoordSf) |>
    ggplot2::StatSfCoordinates$compute_group(coord = ggplot2::CoordSf) ->
australia_state_ref
```

``` r
# layer_county w/ inner join
compute_panel_australia_states <- function(data, scales){

australia_state_ref  |>
    inner_join(data)

}

StatAuStates <- ggproto("StatAuStates",
                        Stat,
                        compute_panel = compute_panel_australia_states,
                        default_aes = aes(label = after_stat(state_name)),
                        required_aes = c("state_name"))

tribble(~state, ~pop,
        "Victoria", 1,
        "Queensland", 2,
        "New South Wales", 3,
        "Western Australia", 4,
        "Northern Territory", 5,
        "Tasmania", 6,
        "South Australia", 7,
        "Australian Capital Territory", 8,
        "Other Territories", 9
        ) |>
ggplot() +
  aes(state_name = state) + 
  geom_sf(stat = StatAuStates) + 
  geom_sf_text(stat = StatAuStates,
               color = "white", 
               check_overlap = T,
               size = 2) + 
  aes(fill = state) +
  theme(panel.background = element_rect(fill = "blue")) + 
  coord_sf(crs = crs_au_states)
```

![](https://i.imgur.com/t9M5vUZ.png)

<sup>Created on 2024-11-18 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
I'm trying to figure out if this approach w/ make_constructor gives you everything you'd want for making a layer like geom_state(), geom_state_label()...  looks probable.

**Reply by EvaMaeRey** [reply]
https://github.com/EvaMaeRey/sf2stat?tab=readme-ov-file#stat_region-and-friends
